### PR TITLE
Implement EnqueueExtensions interface in TopologySpreading scheduling

### PR DESCRIPTION
**Pull request by yuzhiquan**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Implement EnqueueExtensions interface in TopologySpreading scheduling
#### Which issue(s) this PR fixes:

Part of #94009

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/triage accepted
/priority important-longterm
/sig scheduling

### CODE REVIEW

**Comment by yuzhiquan** [comment]
@Huang-Wei 
I run perf test like this
```bash
make test-integration WHAT=./test/integration/scheduler_perf KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-alsologtostderr=false -logtostderr=false -run=^$$ -benchtime=1ns -bench=BenchmarkPerfScheduling/TopologySpreading/500Nodes  -data-items-dir ~/logs/$(git rev-parse --abbrev-ref HEAD)/1000"
```
encounter an fail result like this:
```bash
E0406 15:36:15.876368   15557 scheduler.go:342] "Error updating pod" err="Patch \"http://127.0.0.1:38687/api/v1/namespaces/namespace-1/pods/pod-cgcpl/status\": dial tcp 127.0.0.1:38687: connect: connection refused" pod="namespace-1/pod-cgcpl"
--- FAIL: BenchmarkPerfScheduling/PreferredTopologySpreading/500Nodes
    scheduler_perf_test.go:466: op 1: Error creating pod: Failed to create object with non-retriable error: the server has received too many requests and has asked us to try again later (post pods)
--- FAIL: BenchmarkPerfScheduling/PreferredTopologySpreading
--- FAIL: BenchmarkPerfScheduling
FAIL
exit status 1
FAIL    k8s.io/kubernetes/test/integration/scheduler_perf       163.076s
FAIL
make[1]: *** [Makefile:184：test] 错误 1
!!! [0406 15:36:23] Call tree:
!!! [0406 15:36:23]  1: hack/make-rules/test-integration.sh:97 runTests(...)
+++ [0406 15:36:23] Cleaning up etcd
+++ [0406 15:36:23] Integration test cleanup complete
make: *** [Makefile:203：test-integration] 错误 1

```
any indea on this?

**Comment by chendave** [comment]
This is an known issue,  apply either of below PR and retry, 
- https://github.com/kubernetes/kubernetes/pull/96704
- https://github.com/kubernetes/kubernetes/pull/96696

You will properly get through simply by retrying several times without any change.

File: pkg/scheduler/framework/plugins/podtopologyspread/plugin.go [hunk 1]
```diff
@@ -119,3 +120,13 @@ func (pl *PodTopologySpread) setListers(factory informers.SharedInformerFactory)
 	pl.replicaSets = factory.Apps().V1().ReplicaSets().Lister()
 	pl.statefulSets = factory.Apps().V1().StatefulSets().Lister()
 }
+
+// EventsToRegister returns the possible events that may make a Pod
+// failed by this plugin schedulable.
+func (pl *PodTopologySpread) EventsToRegister() []framework.ClusterEvent {
```
Comment (chendave) [inline pkg/scheduler/framework/plugins/podtopologyspread/plugin.go:126]: You are target `PodTopologySpread` but you claim to address taint toleration?

**Review by chendave — COMMENTED** [review]
pls update the PR title?

File: pkg/scheduler/framework/plugins/podtopologyspread/plugin.go [hunk 2]
```diff
@@ -119,3 +120,13 @@ func (pl *PodTopologySpread) setListers(factory informers.SharedInformerFactory)
...
+	return []framework.ClusterEvent{
+		// pod may be fail by new pod created or pod delete in some node with specified label
+		// And updated node may make pods schedulable.
+		{Resource: framework.Pod, ActionType: framework.Add | framework.Delete | framework.Update},
```
Comment (adtac) [inline pkg/scheduler/framework/plugins/podtopologyspread/plugin.go:130]: new nodes could also make a pod schedulable right?

File: pkg/scheduler/framework/plugins/podtopologyspread/plugin.go [hunk 3]
```diff
@@ -119,3 +120,13 @@ func (pl *PodTopologySpread) setListers(factory informers.SharedInformerFactory)
 	pl.replicaSets = factory.Apps().V1().ReplicaSets().Lister()
 	pl.statefulSets = factory.Apps().V1().StatefulSets().Lister()
 }
+
+// EventsToRegister returns the possible events that may make a Pod
+// failed by this plugin schedulable.
+func (pl *PodTopologySpread) EventsToRegister() []framework.ClusterEvent {
+	return []framework.ClusterEvent{
+		// pod may be fail by new pod created or pod delete in some node with specified label
+		// And updated node may make pods schedulable.
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/plugin.go:129]:
```
// All ActionType includes the following events:
// - Add. An unschedulable Pod may fail due to violating topology spread constraints,
// adding an assigned Pod may make it schedulable.
// - Update. Updating on an existing Pod's labels (e.g., removal) may make
// an unschedulable Pod schedulable.
// - Delete. An unschedulable Pod may fail due to violating an existing Pod's topology spread constraints,
// deleting an existing Pod may make it schedulable.
```

File: pkg/scheduler/framework/plugins/podtopologyspread/plugin.go (hunk 2 above)
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/plugin.go:130]:
```suggestion
		{Resource: framework.Pod, ActionType: framework.All},
```

Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/podtopologyspread/plugin.go:130]: Yes, NodeAdd/NodeDelete/NodeUpdateLable will re-evaluate the Pods spreading, and the unschedulable Pod may become schedulable:

```go
{Resource: framework.Node, ActionType: framework.Add | framework.Delete | framework.UpdateNodeLabel},
```

Comment (yuzhiquan) [reply pkg/scheduler/framework/plugins/podtopologyspread/plugin.go:130]: New node add will with label seted?

Comment (yuzhiquan) [reply pkg/scheduler/framework/plugins/podtopologyspread/plugin.go:130]: Node delete will with label unset, this will affect scheduling.

**Comment by Huang-Wei** [comment]
LGTM. Please squash the commits.

**Pull request merged**